### PR TITLE
CORE-17430 Persist reference states into utxo_transaction_sources table and align with column changes

### PR DESCRIPTION
**Pull request by nkovacsx**
### Overview

1. The `utxo_transaction_sources` table has been modified. See https://github.com/corda/corda-api/pull/1288. This PR contains the changes to align with those DB changes.
2. Reference states were not persisted in `utxo_transaction_sources` until now. Added that persisting logic.
3. Extra: Move `findTransactionIdsAndStatuses` SQL to the provider. We used a raw SQL before.

### Test plan

See https://github.com/corda/corda-api/pull/1288.

### CODE REVIEW

File: components/ledger/ledger-persistence/testing-datamodel/src/main/kotlin/com/example/ledger/testing/datamodel/utxo/UtxoTransactionEntity.kt [hunk 1]
```diff
@@ -48,6 +48,8 @@ data class UtxoTransactionEntity(
         if (!privacySalt.contentEquals(other.privacySalt)) return false
         if (accountId != other.accountId) return false
         if (created != other.created) return false
+        if (status != other.status) return false
```
Comment (relyafi) [inline components/ledger/ledger-persistence/testing-datamodel/src/main/kotlin/com/example/ledger/testing/datamodel/utxo/UtxoTransactionEntity.kt]: Is there a reason we've added these fields as part of these changes? These do not contribute to the uniqueness of the entity so they shouldn't matter. Though I think this applies to a number of the existing fields too.

There's a number of articles on this topic, such as this one: https://thorben-janssen.com/ultimate-guide-to-implementing-equals-and-hashcode-with-hibernate/

File: components/ledger/ledger-persistence/testing-datamodel/src/main/kotlin/com/example/ledger/testing/datamodel/utxo/UtxoTransactionEntity.kt [hunk 2]
```diff
@@ -57,6 +59,8 @@ data class UtxoTransactionEntity(
         result = 31 * result + privacySalt.contentHashCode()
         result = 31 * result + accountId.hashCode()
         result = 31 * result + created.hashCode()
+        result = 31 * result + status.hashCode()
+        result = 31 * result + updated.hashCode()
```
Comment (relyafi) [inline components/ledger/ledger-persistence/testing-datamodel/src/main/kotlin/com/example/ledger/testing/datamodel/utxo/UtxoTransactionEntity.kt]: As above

File: testing/ledger/ledger-hsqldb/src/main/kotlin/net/corda/testing/ledger/utxo/HsqldbUtxoQueryProvider.kt [hunk 3]
```diff
@@ -30,12 +30,24 @@ class HsqldbUtxoQueryProvider @Activate constructor(
...
+                VALUES (x.transaction_id, x.group_idx, x.leaf_idx, x.referenced_state_transaction_id, x.referenced_state_index)"""
+            .trimIndent()
+
     override val persistTransactionComponentLeaf: String
         get() = """
             MERGE INTO {h-schema}utxo_transaction_component AS utc
             USING (VALUES :transactionId, CAST(:groupIndex AS INT), CAST(:leafIndex AS INT), CAST(:data AS VARBINARY(1048576)), :hash)
                 AS x(transaction_id, group_idx, leaf_idx, data, hash)
-            ON x.transaction_id = utc.transaction_id AND x.group_idx = utc.group_idx AND x.leaf_idx = utc.leaf_idx
+            ON x.transaction_id = utc.transaction_id AND x.group_idx = utc.group_idx AND x.leaf_idx = utc.leaf_idx
```
Comment (relyafi) [inline testing/ledger/ledger-hsqldb/src/main/kotlin/net/corda/testing/ledger/utxo/HsqldbUtxoQueryProvider.kt]: Unnecessary whitespace change?

File: components/ledger/ledger-persistence/src/integrationTest/kotlin/net/corda/ledger/persistence/utxo/tests/UtxoLedgerMessageProcessorTests.kt [hunk 4]
```diff
@@ -207,7 +207,7 @@ class UtxoLedgerMessageProcessorTests {
                 listOf("4".toByteArray()),
                 listOf("5".toByteArray()),
                 emptyList(),
-                listOf("7".toByteArray()),
+                emptyList(),
```
Comment (relyafi) [inline components/ledger/ledger-persistence/src/integrationTest/kotlin/net/corda/ledger/persistence/utxo/tests/UtxoLedgerMessageProcessorTests.kt:210]: What is the purpose of this change?

**Review by relyafi — COMMENTED** [review]
A few minor comments, but broadly looks good. As we're not tackling exposing APIs to utilize the data at this point though, having discussed with @blsemo I think we need to create two further tickets off the back of this:

1) Provide a mechanism to migrate reference states created in 5.0.x releases that will not be populated in `utxo_transaction_sources` (or make a statement that 5.0.x data is not supported)
2) Implement APIs for efficient input and reference state retrieval. This ticket would depend on 1)

File: components/ledger/ledger-persistence/testing-datamodel/src/main/kotlin/com/example/ledger/testing/datamodel/utxo/UtxoTransactionEntity.kt (hunk 1 above)
Comment (nkovacsx) [reply components/ledger/ledger-persistence/testing-datamodel/src/main/kotlin/com/example/ledger/testing/datamodel/utxo/UtxoTransactionEntity.kt]: I figured we want to use all the fields to generate the `hashCode()` as it seemed like this is the practice we are following. 
I'm happy to revert this change but that means we need to modify all the other entities because they all follow the same practice, which I'd rather not do as part of this PR.
Also, this is only used in integration tests.

File: components/ledger/ledger-persistence/src/integrationTest/kotlin/net/corda/ledger/persistence/utxo/tests/UtxoLedgerMessageProcessorTests.kt (hunk 4 above)
Comment (nkovacsx) [reply components/ledger/ledger-persistence/src/integrationTest/kotlin/net/corda/ledger/persistence/utxo/tests/UtxoLedgerMessageProcessorTests.kt:210]: Before, we never used the `wrappedWireTransaction.referenceStateRefs` call, which deserialises the reference group:
```
val referenceStateRefs: List<StateRef> by lazy(LazyThreadSafetyMode.PUBLICATION) {
        deserialize(UtxoComponentGroup.REFERENCES)
}
```
so having a non-stateref object here was fine but now we do want to de-serialize and "7" cannot be deserialized as a `StateRef` object.

**Review by vlajos — DISMISSED** [review]

File: components/ledger/ledger-persistence/testing-datamodel/src/main/kotlin/com/example/ledger/testing/datamodel/utxo/UtxoTransactionEntity.kt (hunk 1 above)
Comment (relyafi) [reply components/ledger/ledger-persistence/testing-datamodel/src/main/kotlin/com/example/ledger/testing/datamodel/utxo/UtxoTransactionEntity.kt]: True. But I guess the PR to change it in would have been the data model simplifications one that moved these fields? We've not done anything in this PR that makes this change necessary?

Comment (nkovacsx) [reply components/ledger/ledger-persistence/testing-datamodel/src/main/kotlin/com/example/ledger/testing/datamodel/utxo/UtxoTransactionEntity.kt]: Yeah it should have been done in the previous PR but I just realised I didn't do it. I didn't want to raise a separate PR for it though.

**Review by relyafi — DISMISSED** [review]

**Review by relyafi — DISMISSED** [review]

**Review by vlajos — DISMISSED** [review]

**Review by blsemo — DISMISSED** [review]
LGTM

**Review by vlajos — APPROVED** [review]

**Review by blsemo — APPROVED** [review]

**Review by relyafi — APPROVED** [review]
